File: python/61_序列化二叉树.py
#!/bin/bash/ python3
#-*- coding:utf-8 -*-

# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def Serialize(self, root):
        # write code here
        if root == None:
            return "#"
        
        return str(root.val) + "!" + self.Serialize(root.left) + self.Serialize(root.right)
    
    def my_Deserialize(self, s):
        # 传递列表更加方便
        if s[0] == "#":
            s.pop(0)
            return None
        
        # 获取当前值
        index = s.index("!")
        curr_val = int("".join(s[:index]))
        del s[:index + 1]
        #s = s[index + 1:] // 这是错误的方式
        
        currNode = TreeNode(curr_val)
        currNode.left = self.my_Deserialize(s)
        currNode.right = self.my_Deserialize(s)
        
        return currNode
        
    def Deserialize(self, s):
        # 注意字符串参数是传值的
        if s == "":
            return None
        
        tmp = list(s)
        return self.my_Deserialize(tmp)

File: cpp/61_序列化二叉树.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    std::string ret;
    
    std::string my_Serialize(TreeNode *root) {
        if (root == nullptr) {
            return "#";
        } else {
            // 先序遍历
            return std::to_string(root->val) + "!" + my_Serialize(root->left) + my_Serialize(root->right);
        }
    }
    
    TreeNode* my_Deserialize(char* &str) {
        if (*str == '#') {
            return nullptr;
        }
        
        // 提取当前结点值
        int curr_val = 0;
        while (*str != '!') {
            curr_val = curr_val * 10 + (*str - '0');
            str++;
        }
        
        TreeNode* currNode = new TreeNode(curr_val);
        currNode->left = my_Deserialize(++str);
        currNode->right = my_Deserialize(++str);
        
        return currNode;
    }
    
    char* Serialize(TreeNode *root) {    
        if (root == nullptr) {
            return nullptr;
        }
        
        this->ret = my_Serialize(root);
        return (char*)(ret.c_str());
    }
    
    TreeNode* Deserialize(char *str) {
        if (str == nullptr) {
            return nullptr;
        }
        
        return my_Deserialize(str);
    }
};